Provide a way to handle GraphQL errors

Bogyie opened the pull request:
Motivation:

fixes #4423 

Modifications:

- Add GraphqlErrorHandler
- Add fields errorHandler in GraphqlServiceBuilder, then add method errorHandler when setting errorHandler
- Add fields errorHandler in DefaultGraphqlService, then update constructor
- Add GraphqlErrorHandlerTest

Result:

- Closes #4423 . (If this resolves the issue.)
- You can return different responses depending on the `List<GraphqlError>`

Example:

```java
public interface GraphqlErrorHandler {

     HttpResponse handle(ServiceRequestContext ctx, ExecutionInput input,
                        ExecutionResult result, MediaType negotiatedProduceType, @Nullable Throwable cause);
}

GraphqlErrorHandler errorHandler = 
    new GraphqlErrorHandler() {
        @Override
        @Nullable
        public HttpResponse handle(ServiceRequestContext ctx, ExecutionInput input,
                                             ExecutionResult result, MediaType negotiatedProduceType,
                                             @Nullable Throwable cause) {
            ....
        }
    };

GraphqlService.builder()
              .schemaFile(...)
              .runtimeWiring(...)
              .errorHandler(errorHandler)
              .build();
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=4435) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=4435) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=4435) it.</sub>

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -100,41 +97,9 @@ private HttpResponse execute(ServiceRequestContext ctx, ExecutionInput input, Me
-                                                            .build());
+        return HttpResponse.from(
+                future.handle((executionResult, cause) ->
+                                      errorsHandler.handle(ctx, input, produceType, executionResult, cause)));
```
This is no longer an `errorHandler` because it produces a response even if the result is successful.
How about calling `handle` only if there's an error?
```java
return HttpResponse.from(
        future.handle((executionResult, cause) -> {
            if (cause == null && executionResult.getErrors().isEmpty()) {
                return HttpResponse.ofJson(produceType, executionResult.toSpecification());
            }
            if (executionResult.getData() instanceof Publisher) {
                logger.warn("executionResult.getData() returns a {} that is not supported yet.",
                            executionResult.getData().toString());
                final ExecutionResult error = newExecutionResult(
                        new UnsupportedOperationException("WebSocket is not implemented"));
                return HttpResponse.ofJson(HttpStatus.NOT_IMPLEMENTED, produceType,
                                           error.toSpecification());
            }

            return errorsHandler.handle(ctx, input, produceType, executionResult, cause);
        }));
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,55 @@
+import graphql.GraphQLError;
+
+/**
+ * Error handler which map a GraphQL errors to an {@link HttpResponse}.
```
```suggestion
 * A handler that maps GraphQL errors to an {@link HttpResponse}.
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,55 @@
+public interface GraphqlErrorsHandler {
+
+    /**
+     * Returns a {@link HttpResponse} based on {@link Throwable} or List of {@link GraphQLError}.
```
```suggestion
     * Returns the default {@link GraphqlErrorsHandler}.
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,55 @@
+     * Returns a {@link HttpResponse} based on {@link Throwable} or List of {@link GraphQLError}.
+     */
+    static GraphqlErrorsHandler of() {
+        return GraphqlErrorsHandlers.of(GraphqlErrorsMappingFunction.of());
```
How about creating a default instance and returning it?

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /**
+     * Returns a {@link HttpResponse} based on {@link Throwable} or List of {@link GraphQLError}.
```
How about
```suggestion
     * Returns a {@link GraphqlErrorsHandler} with the {@link GraphqlErrorsMappingFunction} that returns a
     * different {@link HttpStatus} depending on the errors raised in the {@link GraphqlService}.
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,55 @@
+     * @param errorsMappingFunction the function which maps the {@link GraphQLError} to an {@link HttpStatus}.
+     */
+    static GraphqlErrorsHandler of(GraphqlErrorsMappingFunction errorsMappingFunction) {
+        return GraphqlErrorsHandlers.of(errorsMappingFunction);
```
```suggestion
        requireNonNull(errorsMappingFunction, "errorsMappingFunction");
        return GraphqlErrorsHandlers.of(errorsMappingFunction);
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021 LINE Corporation
```
nit: 2022 😉

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    private static HttpResponse toHttpResponse(HttpStatus httpStatus, ExecutionResult executionResult,
+                                               MediaType produceType) {
+        // TODO: When WebSocket is implemented, it should be removed.
+        if (executionResult.getData() instanceof Publisher) {
+            logger.warn("executionResult.getData() returns a {} that is not supported yet.",
+                        executionResult.getData().toString());
+            final ExecutionResult error =
+                    newExecutionResult(new UnsupportedOperationException("WebSocket is not implemented"));
+            return HttpResponse.ofJson(HttpStatus.NOT_IMPLEMENTED, produceType, error.toSpecification());
+        }
```
Now we can remove this. 😉 
```suggestion
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,100 @@
+            }
+            final List<GraphQLError> errors = executionResult.getErrors();
+            final HttpStatus httpStatus = mappingFunction.apply(ctx, input, errors);
+            assert httpStatus != null; // withDefault
```
There's a chance that the `mappingFunction` returns `null` so we should handle it. 
How about adding `GraphqlErrorsMappingFunction.orElse(default)` method and apply it here?

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -230,6 +233,14 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
+     * Sets the {@link GraphqlErrorsHandler}.
+     */
+    public GraphqlServiceBuilder errorsHandler(GraphqlErrorsHandler errorsHandler) {
+        this.errorsHandler = errorsHandler;
```
```suggestion
        this.errorsHandler = requireNonNull(errorsHandler, "errorsHandler");
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -252,7 +263,9 @@ public GraphqlService build() {
             configurer.accept(dataLoaderRegistry);
         }
-        return new DefaultGraphqlService(builder.build(), dataLoaderRegistry, useBlockingTaskExecutor);
+        errorsHandler = errorsHandler != null ? errorsHandler : GraphqlErrorsHandler.of();
```
Let's add `GraphqlErrorsHandler.orElse(another)` method so that we can easily chain the handler:
```java
MyErrorsHandler {

    public HttpResponse handler(...) {
        if (cause instance of MyCause) {
            return myHttpResponse;
        }
        return null; to apply the default handler;
    }
}
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlStatus.java:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2021 LINE Corporation
```
nit: 2022

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlStatus.java:
```diff
@@ -0,0 +1,40 @@
+import graphql.validation.ValidationError;
+
+final class GraphqlStatus {
+    public static HttpStatus graphqlErrorsToHttpStatus(List<GraphQLError> errors) {
```
Let's remove this class and put this method into `GraphqlErrorsHandlers`

minwoox commented on graphql/src/test/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2021 LINE Corporation
```
nit: 2022

minwoox reviewed (commented):
Basically, looks good. Left some suggestions. 😄

bog-y commented:
Hello, @minwoox. It was my first contribution to the open source and I really enjoyed it. I learned a lot from reflecting your advice on the code. Thank you.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,106 @@
+            // invalidate documentation.
+            return HttpStatus.BAD_REQUEST;
+        }
+        return HttpStatus.UNKNOWN;
```
`UNKNOWN` is for internal use only so we should use `SERVICE_UNAVAILABLE` instead.
How about merging `GraphqlErrorsHandler` and `GraphqlErrorsMappingFunction` into `GraphqlErrorsHandler`?
I don't see any particular reason to separate them for now. If we really need the mapping function, we can add it later.

bog-y replied:
I didn't know about `UNKNOWN` and `SERVICE_UNAVAILABLE`, because I haven't written the Http status code in detail yet. I'll fix it right away. Thank you!

The reason for separating Handler and MappingFunction is that I wanted to provide a simple way to convert GraphQL error to Http response. And shamefully, I couldn't think of implementing the orElse method on Handler, so I decided that Handler should have only one processing logic.

Additionally, if MappingFunction is merged with Handler, I would like to reduce the number of Parameters of Handler, is there a good way?

If only `ServiceRequestContext`, `ExecutionInput`, and `List<GraphQLError>` were delivered as you first restricted, the logic already processed in the `execute` would be re-executed in the Handler, so I thought the logic was inefficient.

What are the challenges of implementing the Handler interface as this?
```java
public interface GraphqlErrorsHandler {
    handler(ExecutionResult executionResult, Throwable cause, List<GraphQLError> errors) {
        ...
    }
}
```

minwoox replied:
Adding `ServiceRequestContext` and `ExecutionInput` to the parameter is not for executing the method twice, but for providing more information to the users who customize the functions so that they can use additional information while creating an `HttpResponse`.

> What are the challenges of implementing the Handler interface as this?

How about adding `ServiceRequestContext` and `ExecutionInput` as our current interface does?
```java
public interface GraphqlErrorsHandler {
    handle(ServiceRequestContext, ExecutionInput input, ExecutionResult result, MediaType negotiatedProduceType, @Nullable Throwable cause) {
        ...
    }
}
```
Please note that we don't need to pass `List<GraphQLError>` because we can get it from the `ExecutionResult`.

Bogyie replied:
Thank you for your kind advice. I solved the problem related to Http status, and as you advised, I merged MappingFunction into ErrorsHandler. After modifying the code, I think this is much better, too.

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,90 @@
+
+final class GraphqlErrorsHandlers {
+
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlErrorsHandlers.class);
```
This logger is not used. 😉

minwoox commented on graphql/src/test/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlerTest.java:
```diff
@@ -0,0 +1,112 @@
+        final HttpRequest request = HttpRequest.builder().post("/graphql")
+                                               .content(MediaType.GRAPHQL, "{error}")
+                                               .build();
+        final AggregatedHttpResponse response = BlockingWebClient.of(server.httpUri())
```
ditto

minwoox commented on graphql/src/test/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlerTest.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Test
+    void shouldHandleFooError() {
```
I think foo error doesn't mean anything. 😄
How about renaming this to `handledError` and `unhandledError` for the below method?

Bogyie commented:
Hi, @minwoox.  Sorry for resolving the conversation before committing. 😅
I've fixed all the codes that need to be modified. 

There was a difficult part when I modified it. 
ErrorsHandlers' of mehtod did the same thing as the newly added of method to ErrorsHandled and removed it.
Is this the right choice? Would it have been better to leave a method in ErrorsHandlers?

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,70 @@
+public interface GraphqlErrorsHandler {
+
+    /**
+     * Return an {@link HttpStatus} based on the specified list of {@link GraphQLError}s.
```
```suggestion
     * Return the default {@link GraphqlErrorsHandler}.
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,72 @@
+                return HttpResponse.ofJson(httpStatus, negotiatedProduceType, result.toSpecification());
+            };
+
+    private GraphqlErrorsHandlers() {}
```
Let's put this private constructor at the very bottom. It's sort of convetion. 😉

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,72 @@
+     * Return an {@link HttpStatus} based on the specified list of {@link GraphQLError}s.
+     */
+    private static HttpStatus graphqlErrorsToHttpStatus(List<GraphQLError> errors) {
+        if (errors.isEmpty()) {
```
I think we can remove this check because `errors` it not empty.
How about adding `assert !errors.isEmtpy();`?
Also, I think we can inline this method.

Bogyie replied:
Basically, I rewrite it like this.
```java
final List<GraphQLError> errors = result.getErrors();
assert !errors.isEmpty(); // Already checked in DefaultGraphqlService#execute

HttpStatus httpStatus;
if (errors.stream().anyMatch(ValidationError.class::isInstance)) {
    // The server SHOULD deny execution with a status code of 400 Bad Request for
    // invalidate documentation.
    httpStatus = HttpStatus.BAD_REQUEST;
} else {
    httpStatus = HttpStatus.SERVICE_UNAVAILABLE;
}
```

In case we can be sure that `errors` is not empty, wouldn't it be okay to write it like this?
```java
HttpStatus httpStatus;
if (result.getErrors().stream().anyMatch(ValidationError.class::isInstance)) {
    // The server SHOULD deny execution with a status code of 400 Bad Request for
    // invalidate documentation.
    httpStatus = HttpStatus.BAD_REQUEST;
} else {
    httpStatus = HttpStatus.SERVICE_UNAVAILABLE;
}
```

minwoox replied:
> wouldn't it be okay to write it like this?

Yes, we can do that. 😄 
Let's just add `assert !errors.isEmpty();` so that the reviewer can easily check the errors is not empty.

minwoox reviewed (commented):
> Hi, @minwoox. Sorry for resolving the conversation before committing. 😅

Please, don't worry. 😉 

> GraphqlErrorsHandlers of mehtod do the same thing as the newly added of method to GraphqlErrorsHandler .
So, I removed it. Is this the right choice? 

Yes, we should remove that. 😄 

This is probably the last round from me. Thanks a lot for your effort. 😄

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,72 @@
+
+final class GraphqlErrorsHandlers {
+
+    static final GraphqlErrorsHandler defaultErrorsHandler =
```
`static final` variables are conventionally underscore separated, uppercased.

```suggestion
    static final GraphqlErrorsHandler DEFAULT_ERRORS_HANDLER =
```

jrhee17 replied:
@minwoox pointed out this doesn't necessarily have to be treated as a constant. Let's leave as is 😄

Bogyie replied:
@jrhee17 Got it!

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,72 @@
+            // invalidate documentation.
+            return HttpStatus.BAD_REQUEST;
+        }
+        return HttpStatus.SERVICE_UNAVAILABLE;
```
Q) It seems like the previous behavior was to return a 200 OK.  Is there a reason for this change?

Bogyie replied:
Because Erros in ExecutionResult is not empty.

I think the previous code returned OK in the default processing logic because the user could not handle GraphQlErros and Throwable as needed.

However, users can now handle GraphQlError and Throws. Therefore, if the user wants to implement the same behavior as before, it can be implemented by adding an ErrorHandler that returns OK for all states.

minwoox replied:
Oops, I missed this. Thanks, @jrhee17 for pointing this out. 😄 

@Bogyie I think we have to return 200 OK because we don't want to make a breaking change.
If we change the status code, all clients that receive an error response from this service would get affected, which is a breaking change. 

We can change that only if it's a bug. But returning 200 is not a bug and some of the implementations use 200 status:
https://www.apollographql.com/docs/apollo-server/data/errors/#returning-http-status-codes
(Official spec does not define the response code by the way. 😅 )

Bogyie replied:
@jrhee17 @minwoox Oh... I don't know that. Thank you both for advising me on my mistake. 😅
Got it! I will fix as soon as possiable.

jrhee17 reviewed (commented):
Overall looks good! Left some minor comments 😄

Bogyie commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -100,41 +99,21 @@ private HttpResponse execute(ServiceRequestContext ctx, ExecutionInput input, Me
+                    if (cause == null && executionResult.getErrors().isEmpty()) {
+                        return HttpResponse.ofJson(produceType, executionResult.toSpecification());
+                    }
+                    if (executionResult.getData() instanceof Publisher) {
```
Thanks for yout advice! I'll fix now!

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java:
```diff
@@ -0,0 +1,58 @@
+                if (cause != null) {
+                    // graphQL.executeAsync() returns an error in the executionResult with getErrors().
+                    // Use 500 Internal Server Error because this cause might be unexpected.
+                    final ExecutionResult error = newExecutionResult(cause);
+                    return HttpResponse.ofJson(HttpStatus.INTERNAL_SERVER_ERROR, negotiatedProduceType,
+                                               error.toSpecification());
```
- It would be better to use the original `GraphQLError.toSpecification()` from the `cause` if it is an instance of `GraphQLError`.
```java
if (cause instanceof GraphQLError) {
    ((GraphQLError) cause).toSpecification();
    ...
}
```
- It is not related to this PR but should we need to create `ExecutionResult` although we only use the specification? In addition, `ExecutionResultImpl` is an internal API that we should avoid using if possible. Otherwise, it could break backward compatibility unexpectedly.

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandler.java:
```diff
@@ -0,0 +1,69 @@
+ * A handler that maps {@link ExecutionResult#getErrors()} or the specified {@link Throwable}
+ * to an {@link HttpResponse}.
+ */
+@UnstableApi
+@FunctionalInterface
+public interface GraphqlErrorsHandler {
```
Even though `ExecutionResult` returns a list of errors, it is a bit weird to use Error`s` in the handler. How about simply naming it `GraphqlErrorHandler`?
```suggestion
@UnstableApi
@FunctionalInterface
public interface GraphqlErrorHandler {
```

Bogyie replied:
Is it better to change the name of the interface but keep the explanation of javadoc?

ikhoon replied:
I am OK with keeping the Javadoc.

ikhoon replied:
> Is it right to change all these names?

Right apart from `GraphqlErrorsHandlers`. I prefer to rename `GraphqlErrorsHandlers` to  `DefaultGraphqlErrorsHandler`. 
For example:
```
enum DefaultGraphqlErrorHandler implements GraphqlErrorHandler {
    INSTANCE;

    // Also override nullability with `@Nonnull` so that the default error handler never returns null.
    @Nonnull
    @Override
    HttpResponse handle(...) {
        ...
    }
    
}
```

Bogyie replied:
I've completed the implementation and committed it!

I knew that enum could be used for this purpose, but this is the first time I have actually implemented it. So I'm worried if it's properly implemented. 🤔 Especially, I searched hard, but I don't know if Enum's instance is static final.

https://github.com/line/armeria/blob/58f402c40389a0f18f9e26725fd79d8c4179cb7e/graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java#L32-L34

Anyway, I'm so happy to learn a lot from the advice I received while contributing to this issue that I couldn't learn while studying alone with books. 😀

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -230,6 +233,14 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
     }
 
+    /**
+     * Sets the {@link GraphqlErrorsHandler}.
```
```suggestion
     * Sets the {@link GraphqlErrorsHandler}.
     * If not specified, {@link GraphqlErrorsHandler#of()} is used by default.
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java:
```diff
@@ -0,0 +1,58 @@
+import graphql.ExecutionInput;
+import graphql.ExecutionResult;
+import graphql.validation.ValidationError;
+/*
```
The copyright should be located at the top. 😉

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java:
```diff
@@ -0,0 +1,58 @@
+enum DefaultGraphqlErrorHandler implements GraphqlErrorHandler {
+    INSTANCE;
+
+    @Nonnull
```
We don't need to add `@Nonnull` because `@NonNullByDefault` in `package-info.java` is applied by default.

ikhoon replied:
Don't we need `@Nonnull` since the super.method is annotated with `@Nullable`?

minwoox replied:
I missed that. 😅 Please ignore my comment. 😄

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java:
```diff
@@ -0,0 +1,58 @@
+            ExecutionInput input,
+            ExecutionResult result,
+            MediaType negotiatedProduceType,
+            @Nullable Throwable cause
+    ) {
```
We don't put this parenthesis in a separate line. 😉 
```suggestion
            @Nullable Throwable cause) {
```

minwoox commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorHandler.java:
```diff
@@ -0,0 +1,69 @@
+public interface GraphqlErrorHandler {
+
+    /**
+     * Return the {@link DefaultGraphqlErrorHandler}.
```
```suggestion
     * Return the default {@link GraphqlErrorHandler}.
```
... because `DefaultGraphqlErrorHandler` is a package-private class so we cannot expose it to the Javadoc.

minwoox reviewed (commented):
Left a few more minor comments. 😉
Please also address https://github.com/line/armeria/pull/4435#discussion_r976002194. 😄

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -53,17 +50,20 @@ final class DefaultGraphqlService extends AbstractGraphqlService implements Grap
 
     private final boolean useBlockingTaskExecutor;
 
+    private final GraphqlErrorHandler errorsHandler;
```
```suggestion
    private final GraphqlErrorHandler errorHandler;
```

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -53,17 +50,20 @@ final class DefaultGraphqlService extends AbstractGraphqlService implements Grap
+
     DefaultGraphqlService(GraphQL graphQL, DataLoaderRegistry dataLoaderRegistry,
-                          boolean useBlockingTaskExecutor) {
+                          boolean useBlockingTaskExecutor, GraphqlErrorHandler errorsHandler) {
```
Ditto

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -53,17 +50,20 @@ final class DefaultGraphqlService extends AbstractGraphqlService implements Grap
     protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest req) throws Exception {
-        final MediaType produceType = GraphqlUtil.produceType(ctx.request().headers());
-        if (produceType == null) {
+        final MediaType negotiatedProduceType = GraphqlUtil.produceType(ctx.request().headers());
```
Is there any reason to change this? 🤔

Bogyie replied:
I changed it to the same parameter name of the `GraphqlErrorHandler#handler`.
https://github.com/line/armeria/blob/58f402c40389a0f18f9e26725fd79d8c4179cb7e/graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java#L37-L42

I named the variable negatedProduceType by referring to the JavaDoc of GraphqlUtil.
https://github.com/line/armeria/blob/c087d693a24fa3f0800aaed68c7c5d3657026911/graphql-protocol/src/main/java/com/linecorp/armeria/internal/server/graphql/protocol/GraphqlUtil.java#L34-L35

Bogyie replied:
I changed the logic to not passing the `MediaType` to the `handler` and changed the variable name back to the original one.

minwoox replied:
I recommended it because it's clearer when we pass it to the `GraphqlErrorHandler.handle()`. 😉

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -53,17 +50,20 @@ final class DefaultGraphqlService extends AbstractGraphqlService implements Grap
     }
 
     @Override
     protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest req) throws Exception {
```
I think the `AbstractGraphqlService` should also be changed. WDYT? 🤔 

ex) https://github.com/line/armeria/blob/master/graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java#L71

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorHandler.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    @Nullable
+    HttpResponse handle(ServiceRequestContext ctx, ExecutionInput input,
+                        ExecutionResult result, MediaType negotiatedProduceType, @Nullable Throwable cause);
```
Rather than passing in the `MediaType`, how about getting it from the `ServiceRequestContext`?

Bogyie replied:
Oh, we can get the value from `ServiceRequestContext` when we need it later! I was missing that point. Thank you for the good advice! I'll fix it right away. 😄

minwoox replied:
@heowc I think we need to pass the `negotiatedProduceType` so that the implementation of this interface does not have to process getting the produce type again.
Also, we are going to add a setter for the default produce type. 
https://github.com/graphql/graphql-over-http/pull/215
(We can't just change the default type because it breaks the backward compatilblity.)
If so, we can't just call `GraphqlUtil.produceType` in the `handle` method.

heowc replied:
I am concerned that the argument value increases a lot, but since the `armeria-graphql` module is not yet mature, I think that it may need to be changed in the future. 🤔

minwoox replied:
>  that the argument value increases a lot,

I don't get this. 😆 You mean the produce type will be changed a lot in the future?

minwoox replied:
https://github.com/line/armeria/pull/4451
This is what I have expected that we needed. PTAL also. 😄

Bogyie replied:
Hi @heowc, @minwoox I'm sorry I'm late for personal reasons. I need your advice, because I'm still not good at making decisions

minwoox replied:
Sorry about the late reply. 😉 
Let's keep that as is because this is an unstable API so we can add it later if we need the media type. 😉

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -230,6 +233,15 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
+     * Sets the {@link GraphqlErrorHandler}.
+     * If not specified, {@link GraphqlErrorHandler#of()} is used by default.
+     */
+    public GraphqlServiceBuilder errorsHandler(GraphqlErrorHandler errorsHandler) {
```
```suggestion
    public GraphqlServiceBuilder errorHandler(GraphqlErrorHandler errorHandler) {
```

heowc commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlServiceBuilder.java:
```diff
@@ -230,6 +233,15 @@ public GraphqlServiceBuilder useBlockingTaskExecutor(boolean useBlockingTaskExec
+     * If not specified, {@link GraphqlErrorHandler#of()} is used by default.
+     */
+    public GraphqlServiceBuilder errorsHandler(GraphqlErrorHandler errorsHandler) {
+        this.errorsHandler = requireNonNull(errorsHandler, "errorsHandler");
```
Ditto

heowc commented on graphql/src/test/java/com/linecorp/armeria/server/graphql/GraphqlErrorHandlerTest.java:
```diff
@@ -0,0 +1,102 @@
+            final File graphqlSchemaFile =
+                    new File(getClass().getResource("/test.graphqls").toURI());
+
+            final GraphqlErrorHandler errorsHandler
```
```suggestion
            final GraphqlErrorHandler errorHandler
```

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java:
```diff
@@ -0,0 +1,83 @@
+    private static Map<String, Object> toSpecification(Throwable cause) {
+        requireNonNull(cause, "cause");
+
+        final Map<String, Object> errorMap = new LinkedHashMap<>();
+        errorMap.put("message", cause.getMessage());
+
+        final Map<String, Object> result = new LinkedHashMap<>();
+        result.put("errors", Collections.singletonList(errorMap));
```
Could we use `ImmutableMap` and `ImmutableList` instead?
```java
final Map<String, Object> errorMap = 
    ImmutableMap.of("message", firstNonNull(cause.getMessage(), ""));
...
```

ikhoon replied:
Fixed at [`7794cc4` (#4435)](https://github.com/line/armeria/pull/4435/commits/7794cc453bf1bef91777ef5891c48cee97670eaa)

ikhoon commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java:
```diff
@@ -93,48 +93,42 @@ protected HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest
-        }
-        return HttpResponse.ofJson(produceType, executionResult.toSpecification());
-    }
+    private static Map<String, Object> toSpecification(String message) {
```
How about making this package-private and sharing this with `DefaultGraphqlErrorHandler`?

ikhoon reviewed (approved):
Thanks, @Bogyie! 🙇‍♂️ Left super nits. 😆

jrhee17 commented on graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlErrorHandler.java:
```diff
@@ -0,0 +1,83 @@
+            ExecutionResult result,
+            @Nullable Throwable cause) {
+
+        final MediaType produceType = GraphqlUtil.produceType(ctx.request().headers());
```
Technically, I think a user may modify the header via `ctx#updateRequest`.
Having said this, I think the chances of doing so is low, and it doesn't not make sense to modify the response type depending on the headers at this point.
SGTM

jrhee17 reviewed (approved):
Looks good to me! Please address @ikhoon 's comments! Thanks @Bogyie 🙇 👍 🚀

minwoox reviewed (approved):
Great job and thanks a lot, @Bogyie! 🎉 🎉 🎉

minwoox commented:
@heowc Any other comments? 😊

heowc commented:
> @heowc Any other comments? 😊

I have not received any feedback on this comment. WDYT? 🤔 

https://github.com/line/armeria/pull/4435#discussion_r979190256

minwoox commented:
@heowc Do you mean changing the media type of the response?

heowc commented:
> @heowc Do you mean changing the media type of the response?

Including the media type I thought the user should get a separate desired response.

ex)

media-type: text/plain
body: query is missing

media-type: application/json
body:

```json
{
  "errors": [
     ...
     "messsage": "query is missing"
  ]
}
```

minwoox commented:
That sounds good. Because this PR is merged we need to create another PR. Do you mind sending it? 😆

heowc commented:
> That sounds good. Because this PR is merged we need to create another PR. Do you mind sending it? 😆

let's create an issue for tracking.
